### ISSUE TITLE
Problem with Qn::DataContainerStats arithmetics

### ISSUE THREAD

**Issue by mam-mih-val** [HeavyIonAnalysis/QnTools#26]
Dear @kreisl,
We found a bunch of bugs connected to Qn::DataContainerStats arithmetics.
1. If we process any arithmetics on DataContainer and then try rebin this container, we'll get an exceprion thrown:
std::logic_error
what(): Cannot merge Stats. Please check prior operations.
Aborted (core dumped)
2. If we do any division operation like observed flow divide on resolution, and oserved flow container is multidimensional, we'll get failed any projection axis with
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector of samples is empty.
Of course, the numerator is observable-type, the denumerator is reference-type.

Can you please check this?

**Comment by kreisl** [comment]
Hi, I am having some problems with the solution to the issues, this is related to issue #14.

To 1:
It is clear that we would like to construct an broadly binned observable from finely binned correlations. But conceptually is it obvious how to do the rebinning operation after doing arithmetics with the statistic object? For the statistics there is :
- sum x
- sum (x-xmean)^2
- sum w
- sum w^2
- n

being saved. However, if I perform an arithmetic operation I need to do this on the mean and the error of the mean, which I can calculate from these quantities. Usually if we want to merge to bins we do this on the statistics directly. Since this is not possible after the arithmetic operation, as far as I understand, how do we proceed? Do we just rely on the uncertainty calculated on from subsampling or bootstrapping? If not when do we decide to use one or the other?
If I am missing something crucial please let me know.

To 2:
This shouldn't happen. I will try to reproduce the problem.

**Comment by mam-mih-val** [comment]
> To 1:

Thank you, I've understood, what caused the problem. I think it's okay to rely on on the ucertainty from sampling.

> To 2:
> This shouldn't happen. I will try to reproduce the problem.

I can help you and make a small example of bug reproducing, maybe we can develop a test from the example.

**Comment by mam-mih-val** [comment]
@kreisl,
I've made a program reproducing my problem: https://github.com/mam-mih-val/qn_tools_test
As you can notice, I can only build flow in centrality bins and can't build projections on pt or y axis in narrow bins of other axises (without merging bins). I think it's the main problem. Because I can make rebining of axises before calculation, but I can't look at the correlation projection on one axis in narrow bins of other (choosing only one bin in every left axis).

**Comment by eugene274** [comment]
Confirm 2.
 
### Cast:

Main actors:
 - nom1
 - nom2
 - denom
 - result

Episode actors:
 - nominator
 - ratio
  
###  Prologue:

Arguments appear on the scene. Each has **ten** _Qn::Stats_ entries in the _DataContainer::data__. Their _DataContainer::axes__ are identical and consist of one element: 'integrated' axis [0,1].

### Act 1:

`auto nominator = nom1 * nom2`

Nominator appears from multiplication. But its _DataContainer::data__ is too short - **one** entry only.

### Act 2:

`auto ratio = nominator / denom`

Scene crashes with out of bounds error.

The end.

### UPD1: 
Correlations were produced with rather old version of QnTools (before it became QnTools, actually).

After loading with the new QnTools ALL correlations including correlations with few axes have only one 'integrated' axis with bin edges [0,1] inside.

### UPD2:
It seems the problem is due to Qn::Axis became template class Qn::Axis<T>. Apparently ROOT fails to read _DataContainer::axes__ and leaves it empty.

**Comment by kreisl** [comment]
Concerning the prologue and act1: If the nom1 and nom2 both have only integrated axis they should only be of length one so correspondingly the nominator should only be length one. That is the correct behavior. However, why do nom1 and nom2 have more than one entry in data and have and “integrated” axis. Is the integrated flag set? I do not understand this. Is the problem happening earlier?

**Comment by eugene274** [comment]
Check out UPD2.

I will add schema evolution rule for old version of Qn::Axis accordingly to 
https://root.cern/root/htmldoc/guides/users-guide/ROOTUsersGuide.html#schema-evolution

**Comment by kreisl** [comment]
```
Qn::DataContainerStats a;
Qn::DataContainerStats b;
auto c = a * b
c.IsIntegrated() Gives false
```

It Should be true. During the Apply step in the arithmetic operation. It should not addaxes if they are of type integrated. We need to make a check in AddAxes(std::vector<Qn::Axis>) function to prevent this. Could also cause issues when handling only integrated quantities, which is usually not the case but still.

**Comment by eugene274** [comment]
I unlinked PR, because Mikhail' problem seems to be different.

**Comment by kreisl** [comment]
fixed in #28.
